Add line_height_factor to canvas

jannikhoesch opened the pull request:
The Canvas write_text and measure_text methods accepted multiple lines, but the spacing could not be controlled. By adding a line_height_factor this can now be modified, for example by using a slider, which adjusts the distance between two lines (Visual demonstration in canvas example).

Added to the following backends:
- android
- cocoa
- gtk
- iOS
- winforms

Fixes #2144 

## PR Checklist:
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

Co-authored by: @heroldpls
Co-authored by: @amaekh
Co-authored by: @trighap52
Co-authored by: @karindev

freakboy3742 reviewed (changes requested):
I've only done a quick pass over this PR so far - but on the whole, it looks really solid.  Nice work, everyone!

Two (relatively minor) things stood out in my initial review:

1. `line_height_factor` is a bit verbose as a name for a public-facing API. I'd lean toward using `line_height`, mirroring the [CSS property managing the same property](https://developer.mozilla.org/en-US/docs/Web/CSS/line-height). Right now, it will only handle floats; but in theory, with some extra work, it could be expanded to handle all the values that are legal in CSS (I'm not saying this is necessary for this PR to be accepted - just that would be an area that could be explored in future).
2. This looks to be thoroughly tested in the *core* API, but the backend APIs aren't tested at all. The PR requires a change the to `testbed` app to prove that the backends all work as expected. This will be either a modification to [the multiline_text test](https://github.com/beeware/toga/blob/main/testbed/tests/widgets/test_canvas.py#L706), or a new test following the same general pattern, but exercising multiline text with line height.

(1) is a relatively simple rename; (2) needs a bit more work. The canvas testbed tests are a little annoying to write, because they involve comparison against reference images - and the reference images can be subtly different on every machine depending on specific OS versions, fonts installed, display properties and more. The reference images that exist in the repo are what is produced in CI conditions. 

So - the suggested approach here is to write a testbed test that looks like it's doing the right thing, but with a "known incorrect" reference image. You then push the PR to CI, where it will also fail - but as part of the failed run, the "correct" reference images for each platform should be saved and attached to the CI run. You can then download those images, check that they actually look right, add them to the PR as the "real" images, and re-submit the PR - at which point, it *should* pass.

I hope that explanation makes sense - if it doesn't let me know and I can fill in any gaps.

jannikhoesch commented:
Thanks for your feedback! We are going to have a look at it and try to improve the points you mentioned.

mhsmith commented:
We usually follow CSS for the definitions of properties as well as their names. And CSS defines `line-height` as a ratio of the font size, not the default line height. The default is usually somewhere around 1.2, but it depends on the font.

So I think this argument should accept two options:
* `None` (default) – the current behavior
* A float – a multiple of the font size. 

See #2144 for more details.

jannikhoesch commented:
We are almost done fixing all the suggestions. One thing that does not work is that the android testbed does not save an image when failing in the CI. Is there another way to get this reference image?

freakboy3742 commented:
> We are almost done fixing all the suggestions. One thing that does not work is that the android testbed does not save an image when failing in the CI. Is there another way to get this reference image?

I presume you mean a local run of the CI test suite? I'm not seeing any CI runs on this PR. If so, yes - it is possible... just a little complicated. An Android emulator is essentially an entire VM; so if you want to get files off the device, you need to exfiltrate them from the VM image. 

The easiest way I know of to do this is with the tools provided by Android Studio - if the VM is running, Android Studio's simulator tools will let you inspect the filesystem of the device, and copy those files out to your local machine. You'll be looking for the "data/canvas" folder in the app's files. [This SO answer](https://stackoverflow.com/a/32019146) gives a high level summary of the process. That page also has a version of the same instructions using `adb` from the command line. 

Longer term, it might be nice to wrap those commands up into something Briefcase can do - being able to populate and interrogate user-space files could be a useful capability.

jannikhoesch commented:
It should now be fixed:

- Line height accepts none values
- Uses ratio of font size instead of default line height
- Renamed line_height_factor to line_height
- Added testbed tests

Please review again

freakboy3742 commented on cocoa/src/toga_cocoa/widgets/canvas.py:
```diff
@@ -269,25 +269,28 @@ def _render_string(self, text, font, **kwargs):
+            # descender is a negative number.
+            return ceil(font.native.ascender - font.native.descender)
+        else:
+            return font.native.ascender * line_height
```
Are you sure this should be `ascender`, and not `fontSize`? The two are likely very similar, but I'm not sure they're *required* to be identical.

freakboy3742 commented on gtk/src/toga_gtk/widgets/canvas.py:
```diff
@@ -261,20 +263,25 @@ def _pango_context(self, font):
+            # get_height was added in Pango 1.44, but Debian Buster comes with 1.42.
+            scaled_line_height = ascent + descent
+        else:
+            scaled_line_height = ascent * line_height
```
Again - I'm not sure "ascent" is necessarily the same as the font size.

freakboy3742 commented on iOS/src/toga_iOS/widgets/canvas.py:
```diff
@@ -253,25 +253,28 @@ def _render_string(self, text, font, **kwargs):
+            # descender is a negative number.
+            return ceil(font.native.ascender - font.native.descender)
+        else:
+            return font.native.ascender * line_height
```
Same ascender vs fontSize issue

freakboy3742 commented on winforms/src/toga_winforms/widgets/canvas.py:
```diff
@@ -300,16 +300,24 @@ def reset_transform(self, draw_context, **kwargs):
+        if line_height is None:
+            return font.metric("LineSpacing")
+        else:
+            return font.metric("EmHeight") * line_height
```
Same issue here - but I'm more certain Em height isn't the same as font size.

freakboy3742 replied:
Huh... I thought em-width was variable with the font, but evidently not. Regardless - as you say, we have the actual font size, so we can use that directly.

freakboy3742 commented on changes/2144.feature.rst:
```diff
@@ -0,0 +1 @@
+A line_height argument can now be used to change the line height of multiline text.
```
```suggestion
The line height of multiline text on a Canvas can be configured.
```

freakboy3742 reviewed (changes requested):
Awesome work - Thanks for the updates! 

The code structure and testing is dead on - no feedback to give there. The only issue that is outstanding is the specifics of the font height calculation. You've used the "ascender height" fairly consistently as the base from which to construct the line height; while that value *should* be very close to the font size, my understanding is that the font size isn't *required* to match the ascender size. [This blog post](https://tonsky.me/blog/font-size/) does a decent explanation of font geometry. 

As far as I'm aware, we should be *literally* using the font-size value (in points); that might require other changes to convert locally into PX sizes. In the case of macOS and iOS at least, ascender is already given in points, so switching to font size should be fairly straightforward.

(Of course, if you're able to defend using ascender on some grounds that I've either misunderstood or misinterpreted what is going on, let me know! I know a decent amount about fonts, but I won't claim to be a domain expert. It's possible I've got something wrong here!)

mhsmith commented:
> As far as I'm aware, we should be _literally_ using the font-size value (in points); that might require other changes to convert locally into PX sizes.

I agree.

@jannikhoesch: I've already pointed you at #2144, which has a perfectly clear example of this.

karindev commented:
Thanks for the feedback! The blog post was very helpful, I haven't done much with fonts before and they are a lot more complex than I realized. Some of the backends used points and some pixels for the line height so the actual font size should now be used with the appropriate units. Also modified the change message to match the suggested one. Should be good for one last review!

freakboy3742 commented on winforms/src/toga_winforms/widgets/canvas.py:
```diff
@@ -300,16 +300,25 @@ def reset_transform(self, draw_context, **kwargs):
+            return font.metric("LineSpacing")
+        else:
+            # Get size in CSS pixels
+            return (font.native.SizeInPoints * 96 / 72) * line_height
```
I'm a little surprised this doesn't require a modification based on the displays DPI scale (cf the winforms `Window.scale_font()`) . Can you confirm this still works if your display is in HiDPI mode?

karindev replied:
The code inside the else statement was written to match the implementation of `Font.metric()`. Did some digging in the code and it seems like `Font.metric()` should maybe be changed to call `Widget.scale_out()` instead of scaling by a hardcoded value.

karindev replied:
Made a mistake in my previous comment, meant to say that the code in canvas could be scaled using `Widget.scale_out()`, and `Font.metric()` could scale similarly using `dpi_scale`.

karindev replied:
Maybe `Font.metric()` should return points instead of CSS pixels since pixels depend on the DPI scale?

freakboy3742 replied:
Ah - if it's working in CSS pixels, then the scaling will be happening automatically. Between that and the video, I'm happy this is correct.

freakboy3742 reviewed (changes requested):
This is looking pretty good - there's one final detail that I'd like to confirm on the Winforms backend - it's possible the implementation is correct, but if you can explicitly verify the behavior will be as expected, then I think we'll be good to go!

Trighap52 commented:
I tried to verify the behaviour as far as I understood it. The line height seems to work as intended. Let me know if further detail is needed. Thanks fo the feedback !

https://github.com/user-attachments/assets/a3cdc245-d36c-4a93-87aa-d4263079cef6

freakboy3742 reviewed (approved):
With the 2-way confirmation that the Winforms math is correct (video demo and verification that font metrics are working in CSS pixels), I'm happy with this. 

Thanks for the PR - Great work team!
